File: lib/security/rate-limit.ts
import Database from 'better-sqlite3'

import { sqlite } from '../db'

export interface RateLimitResult {
  allowed: boolean
  remaining: number
  resetAt: number
}

let lastCleanup = 0
const CLEANUP_INTERVAL = 10 * 60 * 1000

export function rateLimitWithDatabase(
  database: Database.Database,
  key: string,
  maxRequests: number,
  windowMs: number,
  now = Date.now(),
): RateLimitResult {
  const transaction = database.transaction(() => {
    if (now - lastCleanup >= CLEANUP_INTERVAL) {
      database.prepare('DELETE FROM rate_limits WHERE reset_at <= ?').run(now)
      lastCleanup = now
    }

    const entry = database.prepare(
      'SELECT count, reset_at AS resetAt FROM rate_limits WHERE key = ?',
    ).get(key) as { count: number; resetAt: number } | undefined

    if (!entry || entry.resetAt <= now) {
      const resetAt = now + windowMs
      database.prepare(`
        INSERT INTO rate_limits (key, count, reset_at) VALUES (?, 1, ?)
        ON CONFLICT(key) DO UPDATE SET count = 1, reset_at = excluded.reset_at
      `).run(key, resetAt)
      return { allowed: true, remaining: Math.max(0, maxRequests - 1), resetAt }
    }

    if (entry.count >= maxRequests) {
      return { allowed: false, remaining: 0, resetAt: entry.resetAt }
    }

    const nextCount = entry.count + 1
    database.prepare('UPDATE rate_limits SET count = ? WHERE key = ?').run(nextCount, key)
    return {
      allowed: true,
      remaining: Math.max(0, maxRequests - nextCount),
      resetAt: entry.resetAt,
    }
  })

  return transaction.immediate()
}

export function rateLimit(key: string, maxRequests: number, windowMs: number): RateLimitResult {
  return rateLimitWithDatabase(sqlite, key, maxRequests, windowMs)
}

export function getRateLimitHeaders(remaining: number, resetAt: number) {
  return {
    'X-RateLimit-Remaining': String(remaining),
    'X-RateLimit-Reset': String(Math.ceil(resetAt / 1000)),
  }
}

File: backend/src/rate_limit.rs
use sqlx::{Row, SqlitePool};

#[derive(Debug, Clone, Copy)]
pub struct RateLimit {
    pub allowed: bool,
    pub remaining: i64,
    pub reset_at: i64,
}

pub async fn check(
    pool: &SqlitePool,
    key: &str,
    max_requests: i64,
    window_ms: i64,
    now: i64,
) -> Result<RateLimit, sqlx::Error> {
    let mut transaction = pool.begin().await?;
    sqlx::query("DELETE FROM rate_limits WHERE reset_at <= ?")
        .bind(now)
        .execute(&mut *transaction)
        .await?;

    let row = sqlx::query("SELECT count, reset_at FROM rate_limits WHERE key = ?")
        .bind(key)
        .fetch_optional(&mut *transaction)
        .await?;

    let result = if let Some(row) = row {
        let count: i64 = row.try_get("count")?;
        let reset_at: i64 = row.try_get("reset_at")?;
        if count >= max_requests {
            RateLimit {
                allowed: false,
                remaining: 0,
                reset_at,
            }
        } else {
            let next_count = count + 1;
            sqlx::query("UPDATE rate_limits SET count = ? WHERE key = ?")
                .bind(next_count)
                .bind(key)
                .execute(&mut *transaction)
                .await?;
            RateLimit {
                allowed: true,
                remaining: (max_requests - next_count).max(0),
                reset_at,
            }
        }
    } else {
        let reset_at = now + window_ms;
        sqlx::query("INSERT INTO rate_limits (key, count, reset_at) VALUES (?, 1, ?)")
            .bind(key)
            .bind(reset_at)
            .execute(&mut *transaction)
            .await?;
        RateLimit {
            allowed: true,
            remaining: (max_requests - 1).max(0),
            reset_at,
        }
    };

    transaction.commit().await?;
    Ok(result)
}
